// @salmonstats3/av5ja/dist/requests/bullet_token.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Transform, plainToInstance } from 'class-transformer';
import { Method } from '../enum/method';
export var BulletToken;
(function (BulletToken) {
    class Request {
        baseURL = 'https://api.lp1.av5ja.srv.nintendo.net/';
        headers;
        method = Method.POST;
        parameters;
        path = 'api/bullet_tokens';
        constructor(token, version) {
            this.headers = {
                Accept: '*/*',
                'Content-Type': 'application/json',
                Origin: 'https://api.lp1.av5ja.srv.nintendo.net',
                'X-GameWebToken': token.raw_value,
                'X-NaCountry': 'US',
                'X-Requested-With': 'com.nintendo.znca',
                'X-Web-View-Ver': version,
            };
        }
        request(response) {
            return plainToInstance(Response, response, { excludeExtraneousValues: false });
        }
    }
    BulletToken.Request = Request;
    class Response {
        bullet_token;
        lang;
        is_noe_country;
    }
    __decorate([
        Transform(({ value }) => value === 'true'),
        __metadata("design:type", Boolean)
    ], Response.prototype, "is_noe_country", void 0);
    BulletToken.Response = Response;
})(BulletToken || (BulletToken = {}));
