// @salmonstats3/av5ja/dist/requests/coral_token.js
import { plainToInstance } from 'class-transformer';
import { Method } from '../enum/method';
export var CoralToken;
(function (CoralToken) {
    class Request {
        baseURL = 'https://api.imink.app/';
        headers;
        method = Method.POST;
        parameters;
        path = 'f';
        constructor(token, hash_method, na_id, coral_user_id, version) {
            this.headers = {
                'Content-Type': 'application/json',
                'User-Agent': 'av5ja/0.2.2',
                'X-znca-Platform': 'Android',
                'X-znca-Version': version,
            };
            this.parameters = {
                coral_user_id: coral_user_id,
                hash_method: hash_method,
                na_id: na_id,
                token: token,
            };
        }
        request(response) {
            return plainToInstance(Response, response, { excludeExtraneousValues: false });
        }
    }
    CoralToken.Request = Request;
    class Response {
        f;
        request_id;
        timestamp;
    }
    CoralToken.Response = Response;
})(CoralToken || (CoralToken = {}));
