// @salmonstats3/av5ja/dist/requests/user_me.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Expose, plainToInstance } from 'class-transformer';
import { Method } from '../enum/method';
export var UserMe;
(function (UserMe) {
    class Request {
        baseURL = 'https://api.accounts.nintendo.com/';
        headers;
        method = Method.GET;
        parameters;
        path = '2.0.0/users/me';
        constructor(token) {
            this.headers = {
                'Accept-Encoding': 'gzip',
                Authorization: `Bearer ${token.raw_value}`,
                Host: 'api.accounts.nintendo.com',
                'User-Agent': 'NASDKAPI; Android',
            };
        }
        request(response) {
            console.log(response);
            return plainToInstance(Response, response, { excludeExtraneousValues: true });
        }
    }
    UserMe.Request = Request;
    class Response {
        nickname;
        language;
        country;
        birthday;
        id;
    }
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "nickname", void 0);
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "language", void 0);
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "country", void 0);
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "birthday", void 0);
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "id", void 0);
    UserMe.Response = Response;
})(UserMe || (UserMe = {}));
