// @salmonstats3/av5ja/dist/enum/coop_enemy_info_id.js
export var CoopEnemyInfoId;
(function (CoopEnemyInfoId) {
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienBomber"] = 4] = "SakelienBomber";
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienCupTwins"] = 5] = "SakelienCupTwins";
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienShield"] = 6] = "SakelienShield";
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienSnake"] = 7] = "SakelienSnake";
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienTower"] = 8] = "SakelienTower";
    CoopEnemyInfoId[CoopEnemyInfoId["Sakediver"] = 9] = "Sakediver";
    CoopEnemyInfoId[CoopEnemyInfoId["Sakerocket"] = 10] = "Sakerocket";
    CoopEnemyInfoId[CoopEnemyInfoId["SakePillar"] = 11] = "SakePillar";
    CoopEnemyInfoId[CoopEnemyInfoId["SakeDolphin"] = 12] = "SakeDolphin";
    CoopEnemyInfoId[CoopEnemyInfoId["SakeArtillery"] = 13] = "SakeArtillery";
    CoopEnemyInfoId[CoopEnemyInfoId["SakeSaucer"] = 14] = "SakeSaucer";
    CoopEnemyInfoId[CoopEnemyInfoId["SakelienGolden"] = 15] = "SakelienGolden";
    CoopEnemyInfoId[CoopEnemyInfoId["Sakedozer"] = 17] = "Sakedozer";
    CoopEnemyInfoId[CoopEnemyInfoId["SakeBigMouth"] = 20] = "SakeBigMouth";
})(CoopEnemyInfoId || (CoopEnemyInfoId = {}));
export var CoopBossInfoId;
(function (CoopBossInfoId) {
    CoopBossInfoId[CoopBossInfoId["SakelienGiant"] = 23] = "SakelienGiant";
    CoopBossInfoId[CoopBossInfoId["SakeRope"] = 24] = "SakeRope";
})(CoopBossInfoId || (CoopBossInfoId = {}));
