// @salmonstats3/av5ja/dist/requests/stats/coop_result.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Expose, plainToInstance } from 'class-transformer';
import { Method } from '../../enum/method';
import { base_url } from '../../utils/env';
import { request } from '../../utils/request';
import 'reflect-metadata';
export var CoopResult;
(function (CoopResult) {
    class Request {
        baseURL = base_url;
        headers = {
            'Content-Type': 'application/json',
        };
        method = Method.POST;
        parameters;
        path = 'v1/results';
        constructor(results) {
            this.parameters = {
                results: JSON.parse(JSON.stringify(results)),
            };
        }
        request(response) {
            return response.map((v) => plainToInstance(Response, v, { excludeExtraneousValues: true }));
        }
    }
    CoopResult.Request = Request;
    class Response {
        result_id;
        schedule_id;
    }
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "result_id", void 0);
    __decorate([
        Expose(),
        __metadata("design:type", String)
    ], Response.prototype, "schedule_id", void 0);
    CoopResult.Response = Response;
})(CoopResult || (CoopResult = {}));
export async function set_coop_history_details(results) {
    return await request(new CoopResult.Request(results));
}
